### PR TITLE
Objects and Array #39517

### PR DESCRIPTION
**Pull request by shashankbaranwal**
- learn about how to do code using javascript
- learn some method about Arrays
- learn some method about Objects
- learn how to validate users with mail
- also , how to give permission with mail

### CODE REVIEW

File: extra/utils/Validation.js [hunk 1]
```diff
@@ -0,0 +1,51 @@
+let users = [
+    {
+        TraineeEmail : 'shashank.baranawal@successive.tech',
```
Comment (namanpar) [inline extra/utils/Validation.js:3]: use camelCase

File: extra/utils/Validation.js [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+        return false;
+    }
+}
+console.log(validateEmail(users[0]["TraineeEmail"]))
+
+function validateUsers(users){
+    let countValid=0;
+    let countInvalid=0;
+    users.forEach(item => {
+        const {TraineeEmail}= item;
```
Comment (namanpar) [inline extra/utils/Validation.js:33]: no need use destructuring twice for same object

**Review by namanpar — CHANGES_REQUESTED** [review]
Use proper message in PR comment
Only raise PR with files related to task
For validation email task - also print the valid and invalid user separately

File: extra/utils/Validation.js (hunk 1 above)
Comment (namanpar) [reply extra/utils/Validation.js:3]: use camelCase for naming keys of object

File: extra/utils/Permissions.js [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    }
+}
+function hasPermission(moduleName, role, permissionType){
+    console.log(permissions[moduleName][permissionType].includes(role));
+};
+    
+hasPermission('getUsers1', 'trainee', 'all');
+hasPermission('getUsers', 'trainee', 'read');
+hasPermission('getUsers2', 'trainee', 'delete');
+hasPermission('getUser2', 'trainee', 'write');
```
Comment (namanpar) [inline extra/utils/Permissions.js:28]: This is giving error. Try to handle it

File: extra/utils/Validation.js (hunk 1 above)
Comment (shashankbaranwal) [reply extra/utils/Validation.js:3]: ohk naman

File: extra/utils/Validation.js [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+    }
+}
+//console.log(validateEmail(users[0]["traineeEmail"]))
+
+function validateUsers(users){
+    let countValid=0;
+    let countInvalid=0;
+    users.forEach(item => {
+        const {traineeEmail,reviewerEmail}= item;
+        if(validateEmail(item.traineeEmail)=== true && validateEmail(item.reviewerEmail)===true){
```
Comment (namanpar) [inline extra/utils/Validation.js:34]: Use values for de-structured object. Used destructuring but still used object dot method to access

File: extra/utils/Permission.js [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+        else if(permissionType=="Delete"){
+            type = Delete.includes(role)
+            return type;
+        }
+    }
+}
+let result = hasPermission(permissions.getUsers,"head-trainer","all");
+console.log(result);
+let result_1 = hasPermission(permissions.getUsers,"trainee","read");
+console.log(result_1);
```
Comment (namanpar) [inline extra/utils/Permission.js:38]: Need to pass the moduleName only not module itself

File: extra/utils/Validation.js [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+let a =[];
+let b =[];
+function validateUsers(users){
+    let countValid=0;
+    let countInvalid=0;
+    users.forEach(item => {
+        const {traineeEmail,reviewerEmail}= item;
+        if(validateEmail(item.traineeEmail)){
+            a.push(traineeEmail);
+            countValid++;
```
Comment (namanpar) [inline extra/utils/Validation.js:37]: Still using dot method to access object values

Comment (namanpar) [reply extra/utils/Validation.js:37]: Redundant access of object values

**Review by namanpar — CHANGES_REQUESTED** [review]

File: extra/utils/rr.js [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+            if(permissionType=="read")
+            {
+                f=read.includes(role)
+                return f;
+            }
+            else if(permissionType=="write"){
+                f=write.includes(role);
+                return f;
+            }
+            else if(permissionType=="Delete"){
```
Comment (namanpar) [inline extra/utils/rr.js:23]: Unnecessary file

File: extra/utils/Permission.js [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+    let temp;
+    function hasPermission(moduleName,role,permissionType)
+    {
+        const {all,read,write,Delete}=moduleName;
+         temp = all.includes(role)
+
+        if(temp==true)
+        {
+            return true
+        }
```
Comment (namanpar) [inline extra/utils/Permission.js:27]: Still passing object instead moduleName

**Review by namanpar — CHANGES_REQUESTED** [review]
Kindly resolve the requested change

File: extra/utils/Permission.js (hunk 8 above)
Comment (shashankbaranwal) [reply extra/utils/Permission.js]: correct all the things .

File: extra/utils/rr.js (hunk 7 above)
Comment (shashankbaranwal) [reply extra/utils/rr.js]: by mistaken this file added

File: extra/utils/Validation.js (hunk 6 above)
Comment (shashankbaranwal) [reply extra/utils/Validation.js]: i corrected all the things

File: extra/utils/Permission.js (hunk 5 above)
Comment (shashankbaranwal) [reply extra/utils/Permission.js]: ya i passsed the moduleName and it running correctly

File: extra/utils/Validation.js (hunk 4 above)
Comment (shashankbaranwal) [reply extra/utils/Validation.js]: ya i resolved the problem.

**Review by namanpar — APPROVED** [review]

**Pull request merged**